### PR TITLE
Stretch functional tests

### PR DESCRIPTION
**Pull request by robotprogrammer22**
Functional tests for stretch converted.

## Description
Converted functional tests for stretch to gtest.

## How Has This Been Tested?
Waiting for Jenkins test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [ ] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [ ] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/base/objs/Process/Process.h [hunk 1]
```diff
@@ -240,9 +240,9 @@ namespace Isis {
       virtual void SetInputCube(Isis::Cube *inCube);
 
 
-      virtual Isis::Cube *SetOutputCube(const QString &parameter);
+      virtual Isis::Cube *SetOutputCube(const QString &parameter, UserInterface *ui=nullptr);
```
Comment (jessemapel) [inline isis/src/base/objs/Process/Process.h]: This functional is overriden in ProcessMosaic, so it should be updated there too.

File: isis/src/base/objs/ProcessImport/ProcessImport.h [hunk 2]
```diff
@@ -199,8 +199,7 @@ namespace Isis {
       // ProcessImport objects and child objects
 
       using Isis::Process::SetOutputCube;  // make parent functions visable
-      Isis::Cube *SetOutputCube(const QString &parameter);
-      Isis::Cube *SetOutputCube(const QString &parameter, UserInterface &ui);
+      Isis::Cube *SetOutputCube(const QString &parameter, UserInterface *ui=nullptr);
```
Comment (jessemapel) [inline isis/src/base/objs/ProcessImport/ProcessImport.h]: Changing this argument from a reference to a pointer is API breaking :(

Comment (Kelvinrr) [reply isis/src/base/objs/ProcessImport/ProcessImport.h]: I'm not sure if `Isis::Cube *SetOutputCube(const QString &parameter, UserInterface &ui);` was part of a release yet, if not, would it still be API breaking? At least, I am fairly confident that method was added very recently as part of another test conversion.

Comment (jessemapel) [reply isis/src/base/objs/ProcessImport/ProcessImport.h]: If it hasn't gone out yet, then it's fine. I'm not sure when that was added.

Comment (Kelvinrr) [reply isis/src/base/objs/ProcessImport/ProcessImport.h]: oof, seems like a week before the last release: https://github.com/USGS-Astrogeology/ISIS3/commit/fefd69936651e05bd2832b1ad4acc0b6b69d3574#diff-80c13406587f8a357946c8464e41d20f56424cd9e810cb3abaa41e3ea9101251 

4.3.0 was tagged Oct 26 and the PR is from Oct 19. 

@robotprogrammer22 This basically means you need to remove this version and use Austin's original version of the function which shouldn't be too hard.

**Comment by Kelvinrr** [comment]
ctests are failing with unable to open errors: 

```
C++ exception with description "**I/O ERROR** Unable to open [/var/folders/qn/565q6qnj7hlft978vwxj99n40000gp/T/qt_temp-PNZSPh/default.cub3?|?]." thrown in the test body.
```

Looks like something is munging the paths, this looks like Isis is reading after the end of an array somewhere.

File: isis/src/base/objs/UserInterface/UserInterface.cpp [hunk 3]
```diff
@@ -405,22 +405,22 @@ namespace Isis {
...
       args.prepend("someapp");
     }
 
     c_args = (char**)malloc(sizeof(char*)*args.size());
-    
-    for (size_t i = 0; i < args.size(); i++) {
-      c_args[i] = (char*)malloc(sizeof(char)*args[i].size()+1);
+
+    for (int i = 0; i < args.size(); i++) {
+      c_args[i] = (char*)malloc(sizeof(char)*args[i].size());
```
Comment (Kelvinrr) [inline isis/src/base/objs/UserInterface/UserInterface.cpp]: The +1 one needs to put back, this is why the tests are failing. The +1 makes sure the null terminator is copied over when creating the c-string. 

see: https://www.tutorialspoint.com/what-is-a-null-terminated-string-in-c-cplusplus

File: isis/src/base/apps/stretch/stretch_app.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+#include "PvlGroup.h"
+#include "PvlKeyword.h"
+#include "UserInterface.h"
+
+namespace Isis {
+  extern void stretch(UserInterface &ui, Pvl *log=nullptr);
+
+  extern void stretch(Cube *inCube, QString &pairs, UserInterface &ui, Pvl *log=nullptr);
+
+  //extern void stretch(Cube *inCube, UserInterface &ui, Pvl *log);
```
Comment (AustinSanders) [inline isis/src/base/apps/stretch/stretch_app.h]: Should this be removed or should we add a matching function in the source file?

File: isis/src/base/objs/UserInterface/UserInterface.cpp [hunk 5]
```diff
@@ -405,22 +405,22 @@ namespace Isis {
...
+
     if (ignoreAppName) {
       args.prepend("someapp");
     }
 
     c_args = (char**)malloc(sizeof(char*)*args.size());
-    
-    for (size_t i = 0; i < args.size(); i++) {
+
+    for (int i = 0; i < args.size(); i++) {
```
Comment (AustinSanders) [inline isis/src/base/objs/UserInterface/UserInterface.cpp:415]: What's the rationale for changing this to int?

File: isis/tests/FunctionalTestsStretch.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+using namespace Isis;
+
+static QString STRETCH_XML = FileName("$ISISROOT/bin/xml/stretch.xml").expanded();
+
+// case 1
+TEST_F(SpecialSmallCube, FunctionalTestStretchDefault) {
+  QTemporaryDir prefix;
+  QString outCubeFileName = prefix.path() + "/outTemp.cub";
+  //QVector<QString> args = {"to="+outCubeFileName, "pairs=0:255 255:0", "null=500", "lis=700", "lrs=800", "his=900", "hrs=1000"};
```
Comment (AustinSanders) [inline isis/tests/FunctionalTestsStretch.cpp]: Is this something that can be removed?

File: isis/tests/FunctionalTestsStretch.cpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  QVector<QString> args = {"to="+outCubeFileName, "null=500", "lis=700", "lrs=800", "his=900", "hrs=1000"};
+  QString pairs = "0:255 255:0";
+
+  UserInterface options(STRETCH_XML, args);
+  Pvl log;
+  try {
+    //appfoo_func(options);
```
Comment (AustinSanders) [inline isis/tests/FunctionalTestsStretch.cpp]: We should remove the template code.

File: isis/tests/FunctionalTestsStretch.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  EXPECT_DOUBLE_EQ(oCubeStats->Sum(), 50525);
+  EXPECT_DOUBLE_EQ(oCubeStats->ValidPixels(), 100);
+  EXPECT_DOUBLE_EQ(oCubeStats->StandardDeviation(), 302.16673352386897);
+}
+
+// case 2, changes other special pixels to other special pixels
+TEST_F(SpecialSmallCube, FunctionalTestStretchSwitchSpecial) {
+  QTemporaryDir prefix;
+  QString outCubeFileName = prefix.path() + "/outTemp.cub";
+  //QVector<QString> args = {"to="+outCubeFileName, "pairs=0:255 255:0", "null=hrs", "lis=NULL", "lrs=Lis", "his=lRs", "hrs=HiS"};
```
Comment (AustinSanders) [inline isis/tests/FunctionalTestsStretch.cpp]: Can we get rid of this commented out line?

File: isis/tests/FunctionalTestsStretch.cpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  QVector<QString> args = {"to="+outCubeFileName, "null=hrs", "lis=NULL", "lrs=Lis", "his=lRs", "hrs=HiS"};
+  QString pairs = "0:255 255:0";
+
+  UserInterface options(STRETCH_XML, args);
+  Pvl log;
+  try {
+    //appfoo_func(options);
```
Comment (AustinSanders) [inline isis/tests/FunctionalTestsStretch.cpp]: More template code

**Review by AustinSanders — COMMENTED** [review]
Looks good overall -- just a few questions + some leftover commented out code

File: isis/src/base/objs/UserInterface/UserInterface.cpp (hunk 5 above)
Comment (robotprogrammer22) [reply isis/src/base/objs/UserInterface/UserInterface.cpp:415]: I was getting some warnings about it not being an int during the build.

**Comment by Kelvinrr** [comment]
@AustinSanders do you think this is good now?

**Review by AustinSanders — APPROVED** [review]

**Pull request merged**